File: _submissions/18UCS045_charan/DAY1/Matrix Median.cpp
int Solution::findMedian(vector<vector<int> > &A) {
    int n=A.size();
    int m=A[0].size();
    int min=INT_MAX;int max=INT_MIN;
    for(int i=0;i<n;i++)
    {
        if(min>A[i][0])
        {
            min=A[i][0];
        }
        if(max<A[i][m-1])
        {
            max=A[i][m-1];
        }
    }
    int x=(n*m+1)/2;
    while(min<max)
    {
        int c=0;
        int mid=min+(max-min)/2;
        for(int i=0;i<n;i++)
        {
        c+=upper_bound(A[i].begin(),A[i].end(),mid)-A[i].begin();
        }
        if(c<x)
        {
            min=mid+1;
        }
        else
        {
            max=mid;
        }
        
    }
    return min;
}

File: _submissions/19UCS159_Tharun/Day1/matrixMedian.java
public class matrixMedian {
  public int findMedian(int[][] A) {
    int n = A.length, m = A[0].length;
    int min = Integer.MAX_VALUE, max = Integer.MIN_VALUE;
    for (int row = 0; row < n; row++) {
      min = Math.min(min, A[row][0]);
      max = Math.max(max, A[row][m - 1]);
    }
    // binary search for the answer
    int l = min, r = max;
    while (l <= r) {
      int mid = l + (r - l) / 2;
      int required = (n * m + 1) / 2;
      int found = 0, searchMax = 0, searchMaxFreq = 0;
      for (int row = 0; row < n; row++) {
        int search = binarySearch(A[row], mid); // returns i such that a[0..i]<=mid
        if (search != -1) {
          found += search + 1;
          if (A[row][search] > searchMax) {
            searchMax = A[row][search];
          }
        }
      }
      for (int row = 0; row < n; row++) {
        int first = firstOccurance(A[row], searchMax);
        if (first != -1) {
          searchMaxFreq += (binarySearch(A[row], searchMax) - first + 1);
        }
      }
      if (found == required) {
        return searchMax;
      }
      if (found > required && found - searchMaxFreq < required) {
        return searchMax;
      }
      if (found > required) {
        r = mid - 1;
      } else {
        l = mid + 1;
      }
    }
    return -1;
  }

  private int binarySearch(int[] arr, int x) {
    int l = 0, r = arr.length - 1, ans = -1;
    while (l <= r) {
      int mid = l + (r - l) / 2;
      if (arr[mid] > x) {
        r = mid - 1;
      } else {
        ans = mid;
        l = mid + 1;
      }
    }
    return ans;
  }

  private int firstOccurance(int[] arr, int x) {
    int l = 0, r = arr.length - 1, ans = -1;
    while (l <= r) {
      int mid = l + (r - l) / 2;
      if (arr[mid] == x) {
        ans = mid;
        r = mid - 1;
        continue;
      }
      if (arr[mid] > x) {
        r = mid - 1;
      } else {
        l = mid + 1;
      }
    }
    return ans;
  }

}
